src/resolvelib/resolvers.py

```diff
@@ -98,11 +98,7 @@ def merged_with(self, provider, requirement, parent):
         """
         infos = list(self.information)
         infos.append(RequirementInformation(requirement, parent))
-        candidates = [
-            c
-            for c in self.candidates
-            if provider.is_satisfied_by(requirement, c)
-        ]
+        candidates = list(provider.iter_matches([r for r, _ in infos]))
```

Comment:
It bothers me a bit that we always call list on `iter_matches`. I like the idea that a provider can yield results from `iter_matches` if it makes sense to do so, but the way we describe the method, I suspect people will thing that if they have built a list, they need to do

```python
for candidate in return_list:
    yield candidate
```

when actually this just means we break up and then rebuild the list 🙁 

I don't know what the term is for a function that returns "a list or something like an iterable or an iterator that can be converted to a list"... Surely there must be one? And what would be the type of such a beast (if we were using type annotations)? I feel like the typing people would know how to answer this.

Reply:
I’m torn on this. On one hand it is indeed typical for a provider to form a list at some point anyway, but there are also instances (e.g. pip’s!) where it’s easier to implement an iterator. And I also have some potential optimisation ideas to avoid repeating `list()` calls at some point (#39), so I want to avoid the additional `list()` call in the provider as well.

Maybe I’ll allow returning *either* an iterator or sequence. Reading [`collections.abc`](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes), it’s possible to check `isinstance(candidates, Sequence)` to determine what is what. And we’ll need to rename the function again because it’s no longer returning an iterator.

Reply:
Cool. That's what I was trying to suggest. (And looking at the commit text, yes, "iterable" was the term I couldn't remember, doh).